migrations/capcons/linkmigration.go

```diff
@@ -0,0 +1,327 @@
+	return "LinkMigration"
+}
+
+func (m *LinkMigration) Migrate(
+	addressPath interpreter.AddressPath,
+	value interpreter.Value,
+	inter *interpreter.Interpreter,
+) interpreter.Value {
+
+	pathDomain := addressPath.Path.Domain
+	if pathDomain != common.PathDomainPublic &&
+		pathDomain != common.PathDomainPrivate {
+
+		return nil
+	}
+
+	accountAddress := addressPath.Address
+	pathValue := addressPath.Path
+
+	reporter := m.Reporter
+	accountIDGenerator := m.AccountIDGenerator
+
+	locationRange := interpreter.EmptyLocationRange
+
+	var borrowStaticType *interpreter.ReferenceStaticType
+
+	switch readValue := value.(type) {
+	case *interpreter.CapabilityValue:
+		// Already migrated
+		return nil
+
+	case interpreter.PathLinkValue: //nolint:staticcheck
+		var ok bool
+		borrowStaticType, ok = readValue.Type.(*interpreter.ReferenceStaticType)
+		if !ok {
+			panic(errors.NewUnreachableError())
+		}
+
+	case interpreter.AccountLinkValue: //nolint:staticcheck
+		borrowStaticType = interpreter.NewReferenceStaticType(
+			nil,
+			interpreter.FullyEntitledAccountAccess,
+			interpreter.PrimitiveStaticTypeAccount,
+		)
+
+	default:
+		panic(errors.NewUnreachableError())
+	}
+
+	borrowType, ok := inter.MustConvertStaticToSemaType(borrowStaticType).(*sema.ReferenceType)
+	if !ok {
+		panic(errors.NewUnreachableError())
+	}
+
+	// Get target
+
+	target, _, err := m.getPathCapabilityFinalTarget(
+		inter,
+		accountAddress,
+		pathValue,
+		// TODO:
+		// Use top-most type to follow link all the way to final target
+		&sema.ReferenceType{
+			Authorization: sema.UnauthorizedAccess,
+			Type:          sema.AnyType,
+		},
+	)
+	if err != nil {
+		var cyclicLinkErr CyclicLinkError
+		if goerrors.As(err, &cyclicLinkErr) {
+			if reporter != nil {
+				reporter.CyclicLink(cyclicLinkErr)
+			}
+
+			// TODO: really leave as-is? or still convert?
+			return nil
+		}
+
+		panic(err)
+	}
+
+	// Issue appropriate capability controller
+
+	var capabilityID interpreter.UInt64Value
+
+	switch target := target.(type) {
+	case nil:
+		if reporter != nil {
+			reporter.MissingTarget(addressPath)
+		}
+
+		// TODO: really leave as-is? or still convert?
+		return nil
+
+	case pathCapabilityTarget:
+
+		targetPath := interpreter.PathValue(target)
+
+		capabilityID, _ = stdlib.IssueStorageCapabilityController(
+			inter,
+			locationRange,
+			accountIDGenerator,
+			accountAddress,
+			borrowType,
+			targetPath,
+		)
+
+	case accountCapabilityTarget:
+		capabilityID, _ = stdlib.IssueAccountCapabilityController(
+			inter,
+			locationRange,
+			accountIDGenerator,
+			accountAddress,
+			borrowType,
+		)
+
+	default:
+		panic(errors.NewUnreachableError())
+	}
+
+	// Record new capability ID in source path mapping.
+	// The mapping is used later for migrating path capabilities to ID capabilities,
+	// see CapabilityMigration.
+
+	m.CapabilityIDs[addressPath] = capabilityID
+
+	if reporter != nil {
+		reporter.MigratedLink(addressPath, capabilityID)
+	}
```

Comment:
We have a lot of cases throughout this PR where we see this pattern: `if reporter != nil` then do something with the reporter. Can we move the `nil` check into the `MigratedLink` method itself (just return immediately) to cut down on some redundancy?

Reply:
The check is needed because the reporter is optional. How would we move it _into_ the reporter code itself?

Reply:
If I understand correctly pointer methods in Go can have nil receivers. If we move the nil check to the method we can avoid duplicating it everywhere the method is called.

Reply:
That only works for pointer receivers, but the reporter type here is an interface, so AFAICS this isn't possible, but please let me know if you see a way, happy to improve this

Reply:
Ah okay, that's fine then.